libqtile/widget/graph.py

```diff
@@ -91,34 +92,46 @@ def graphwidth(self):
     def graphheight(self):
         return self.bar.height - self.margin_y * 2 - self.border_width * 2
 
-    def draw_box(self, x, y, values):
-        step = int(self.graphwidth / float(self.samples))
+    @property
+    def step(self):
+        return self.graphwidth / float(self.samples)
+
+    def _for_each_step(self, values):
+        for index, val in enumerate(itertools.islice(
+            values,
+            max(int(-(self.graphwidth / self.step) + len(values)), 0),
+            len(values),
+        )):
+            yield index, val
+
+    def _prepare_context(self):
+        self.drawer.ctx.set_line_join(cairocffi.LINE_JOIN_ROUND)
         self.drawer.set_source_rgb(self.graph_color)
-        for val in values:
+        self.drawer.ctx.set_antialias(cairocffi.ANTIALIAS_NONE)
+        self.drawer.ctx.set_operator(cairocffi.OPERATOR_SOURCE)
+        self.drawer.ctx.set_line_width(self.line_width)
+
+    def draw_box(self, x, y, values):
+        self._prepare_context()
```

Comment:
I'll test the PR later today but am just curious whether the `_prepare_context` method needs to be called each time in the relevant `draw_` method, or whether we could just call it once in `_configure`.

Reply:
each time... actually, we should be creating fresh contexts for each draw too but :shrug:

If the drawer is resized or otherwise modified, the context is recreated and loses it's state.

Reply:
Great. Thanks for explaining.